src/ui/widgets/tui_dirlist_detailed.rs

```diff
@@ -80,113 +80,197 @@ fn print_entry(
+        String::from("")
+    } else if stem.is_empty() || extension.is_empty() {
+        let full = format!("{}{}", stem, extension);
+        let mut truncated = truncate_by_char(&full, drawing_width - 1);
+        truncated.push_str(ELLIPSIS);
+        truncated
+    } else {
         let ext_width = extension.width();
-        buf.set_stringn(x, y, extension, drawing_width, style);
         if ext_width > drawing_width {
-            buf.set_string(x + drawing_width as u16 - 1, y, ELLIPSIS, style);
-        }
-    } else if extension.is_empty() {
-        let stem_width = stem.width();
-        buf.set_stringn(x, y, stem, drawing_width, style);
-        if stem_width > drawing_width {
-            buf.set_string(x + drawing_width as u16 - 1, y, ELLIPSIS, style);
-        }
-    } else {
-        if stem.is_empty() {
-            let ext_width = extension.width();
-            buf.set_stringn(x, y, extension, drawing_width, style);
-            if ext_width > drawing_width {
-                buf.set_string(x + drawing_width as u16 - 1, y, ELLIPSIS, style);
-            }
-        } else if extension.is_empty() {
-            let stem_width = stem.width();
-            buf.set_stringn(x, y, stem, drawing_width, style);
-            if stem_width > drawing_width {
-                buf.set_string(x + drawing_width as u16 - 1, y, ELLIPSIS, style);
-            }
+            // file ext does not fit
+            ELLIPSIS.to_string()
+        } else if ext_width == drawing_width {
+            extension.to_string().replacen('.', ELLIPSIS, 1)
         } else {
-            let stem_width = stem.width();
-            let ext_width = extension.width();
-            if stem_width + ext_width <= drawing_width {
-                // file stem and extension fits
-                buf.set_stringn(x, y, stem, drawing_width, style);
-                buf.set_string(x + stem_width as u16, y, extension, style);
-            } else if ext_width > drawing_width {
-                // file ext does not fit
-                buf.set_string(x, y, ELLIPSIS, style);
-            } else {
-                buf.set_stringn(x, y, stem, drawing_width, style);
-                // file ext fits, but file stem does not
-                let ext_start_idx = if drawing_width < ext_width {
-                    0
-                } else {
-                    (drawing_width - ext_width) as u16
-                };
-                buf.set_string(x + ext_start_idx, y, extension, style);
-                let ext_start_idx = if ext_start_idx > 0 {
-                    ext_start_idx - 1
-                } else {
-                    0
-                };
-                buf.set_string(x + ext_start_idx, y, ELLIPSIS, style);
-            }
+            let stem_width = drawing_width - ext_width;
+            let truncated_stem = truncate_by_char(&stem.to_string(), stem_width - 1);
+            format!("{}{}{}", truncated_stem, ELLIPSIS, extension)
         }
     }
 }
+
```

Comment:
I would prefer if the tests were in their own separate file. In this case, something like, `src/ui/widgets/tui_dirlist_detailed_test.rs` or something

Reply:
The [rust book says](https://doc.rust-lang.org/stable/book/ch11-03-test-organization.html#unit-tests) “The convention is to create a module named tests in each file to contain the test functions and to annotate the module with cfg(test).”

When first checking how to do unit tests in Rust, I found this also a bit... special. However, after some thoughts it somehow makes sense and it seems to be the Rust convention.

Anyway, if you want, I can move them. But maybe spend a second though on this first. ;)

Reply:
Yea, we can have a test module instead :)

Reply:
Hm, so that means that the test modules must be listed in the `mod.rs` and that also means that we are **forced to make private functions public, just to write unit tests**, even if it doesn't make sense to have them declared public. Are you really, really sure? What's the reason behind this? (I haven't thought about this privacy-problem myself before...)

I had just started to rewrite that, but I'm really feeling like doing something very wrong here. Making functions public which should be hidden in their module gives me a severe stomach ache...

Reply:
Oh I see. I didn't account for private functions and stuff.

Okay, we can leave it as it is since its the convention stated by the Rust book.